src/IB/CIBFEMethod.cpp

```diff
@@ -1358,7 +1358,7 @@ CIBFEMethod::computeCOMOfStructure(Eigen::Vector3d& center_of_mass, EquationSyst
     const std::vector<std::vector<double> >& phi = fe->get_phi();
 
     // Extract the nodal coordinates.
-    PetscVector<double>& X_petsc = dynamic_cast<PetscVector<double>&>(*X_system.current_local_solution.get());
+    auto& X_petsc = dynamic_cast<PetscVector<double>&>(*X_system.current_local_solution.get());
     /*if (!X_petsc.closed())*/ X_petsc.close();
```

Comment:
I believe the answer is no; `auto` can be a pointer type but we have to write `auto &` to get a reference.

Reply:
The statement 

`auto& X_petsc = dynamic_cast<PetscVector<double>&>(*X_system.current_local_solution.get());`

would translate to

`PetscVector<double>& X_petsc = dynamic_cast<PetscVector<double>&>(*X_system.current_local_solution.get());`

and not

`PetscVector<double>&& X_petsc = dynamic_cast<PetscVector<double>&>(*X_system.current_local_solution.get());`

Is that right?

Reply:
Thats tricky. Let me write a little test code to answer this.

Reply:
It is a reference type and not an lvalue reference: i.e., it translates into 'what we want' (`PetscVector<double>& X_petsc`). I don't have my copy of Effective Modern C++ handy (which I know explains this rule clearly) but the following code demonstrates:

```cpp
#include <iostream>
#include <type_traits>

struct Test
{
  int x = 42;

  Test()
  {
    std::cout << "Test default ctor\n";
  }

  Test(const Test &t)
  {
    std::cout << "Test copy ctor\n";
  }

  const Test &
  get_current() const
  {
    return *this;
  }
};

int main()
{
  Test test;

  std::cout << "test with auto&:\n";
  auto& other = test.get_current();
  static_assert(std::is_same<decltype(other), const Test &>::value,
                "types should be equal");
  std::cout << "test with auto:\n";
  auto other2 = test.get_current();
  static_assert(std::is_same<decltype(other2), Test>::value,
                "types should be equal");
}
```

since the static assertions pass we know that the types are equal. The console output is also what we expect.

Reply:
Hmmn, OK. Never finding the time to update my C++ skills 😟